### ISSUE TITLE
Add `start_dim` and `end_dim` functionality for common reduction operations.

### ISSUE THREAD

**Issue by selflein** [pytorch/pytorch#54766]
## 🚀 Feature
Add `start_dim` and `end_dim` for common reduction operations such as `torch.sum`, `torch.mean`, etc. which applies the operation to dimensions [start_dim, end_dim). This would work similar to `torch.flatten(input, start_dim=0, end_dim=-1)`.

## Motivation

I often find myself adding helper functions like this
```
def sum_except_batch(x, num_batch_dims=1):
    """Sums all elements of `x` except for the first `num_batch_dims` dimensions."""
    reduce_dims = list(range(num_batch_dims, x.ndimension()))
    return torch.sum(x, dim=reduce_dims)
```
where I explicitly have to deal with shapes in order to reduce over multiple dimensions.

## Pitch

Add additional signature in addition to the existing one for reduction operations.
E.g. for `torch.sum` there is
```
torch.sum(input, dim, keepdim=False, *, dtype=None)
```
This issues proposes adding a further signature
```
torch.sum(input, start_dim=0, end_dim=-1, keepdim=False, *, dtype=None)
```

**Comment by heitorschueroff** [comment]
Adding support for reducing over multiple dimensions is definitely a welcome addition to PyTorch and would make us more compatible with NumPy. However, a solution for this should be generalized and not at an operator level.

Also, to stay compatible with NumPy, the solution would be to make the `dim` parameter accept a single dimension or a tuple of dimensions to reduce over. This also has the advantage of allowing reductions over non-contiguous dimensions.

**Comment by vadimkantorov** [comment]
> Also, to stay compatible with NumPy, the solution would be to make the `dim` parameter accept a single dimension or a tuple of dimensions to reduce over.

This is already the case, you can already pass in `dim = (dim1, dim2)` for many reduction functions

**Comment by selflein** [comment]
True.  The proposed feature is for convenience when dealing with tensors with a lot of dimensions or varying number of dimensions since then keeping track of dimensions introduces a lot of clutter and reduces readability.

However, I have found a nicer alternative to, e.g., code like this
```
def sum_except_batch(x, num_batch_dims=1):
    """Sums all elements of `x` except for the first `num_batch_dims` dimensions."""
    reduce_dims = list(range(num_batch_dims, x.ndimension()))
    return torch.sum(x, dim=reduce_dims)
```

One can use `flatten` to implement something along the lines of the proposed feature
```
x.flatten(start_dim=num_batch_dims).sum(-1)
```

In general it seems, one just can use `x.flatten(start_dim=start_dim, end_dim=end_dim).op(start_dim)` where `op` is a reduction operation to implement reduction of dimensions from `start_dim` to `end_dim`.
